### PR TITLE
bpf: Allow IP fragments via ingress L3 policy

### PR DESCRIPTION
**Pull request by joestringer**
Previously we just gated the IP fragments check on whether policy was
enabled, ie if policy is enabled, then drop all IP fragments. An
upcoming PR (#5294) plans to get rid of those POLICY_*GRESS defines,
which would otherwise cause this code to *always* drop fragments. It's
more correct to check L3-based policy in this case. Move the check into
the ingress policy lookup code.

We don't support IPv4 message reassembly, so if we need to apply L4
policy then we can't ensure that we know the port for an IP fragment.
Therefore, if the packet is a fragment, skip the lookups for L4 policy.
If the L3 policy doesn't allow the traffic, then drop it with the reason
DROP_FRAG_NOSUPPORT.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/5321)

### CODE REVIEW

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
Note that this only makes this more lenient for IPv4. IPv6 still drops all packets that are fragments of a larger IP message: 

https://github.com/cilium/cilium/blob/f8ccf6477f80dbe7c6b83ce03840adec3444bd5a/bpf/bpf_lxc.c#L757-L759

https://github.com/cilium/cilium/blob/8cf71581ae58f6190e7d9401c277b4378aa605aa/bpf/lib/ipv6.h#L61-L74

File: bpf/lib/policy.h [hunk 1]
```diff
@@ -48,10 +48,14 @@ static inline bool identity_is_reserved(__u32 identity)
 	return identity < HEALTH_ID;
 }
 
+#ifndef HAVE_L4_POLICY
+#define HAVE_L4_POLICY 0
```
Comment (ianvernon) [inline bpf/lib/policy.h:52]: Why not have this in `pkg/endpoint/bpf.go`? Not sure if this is a dumb question or not, as BPF-land is still a bit foreign to me.

Comment (joestringer) [reply bpf/lib/policy.h:52]: Good idea. I'll address this with any other feedback I might get.

File: bpf/lib/policy.h [hunk 2]
```diff
@@ -66,18 +70,18 @@ __policy_can_access(void *map, struct __sk_buff *skb, __u32 identity,
...
-	policy = map_lookup_elem(map, &key);
-	if (likely(policy)) {
-		cilium_dbg3(skb, DBG_L4_CREATE, identity, SECLABEL,
-			    dport << 16 | proto);
-
-		/* FIXME: Use per cpu counters */
-		__sync_fetch_and_add(&policy->packets, 1);
-		__sync_fetch_and_add(&policy->bytes, skb->len);
-		goto get_proxy_port;
+	if (HAVE_L4_POLICY && !is_fragment) {
```
Comment (tgraf) [inline bpf/lib/policy.h:73]: Maybe we should get rid of `HAVE_L4_POLICY` entirely?

Comment (joestringer) [reply bpf/lib/policy.h:73]: I expect that this will be a subsequent follow-up given that it's one of the causes for recompilation at the moment.

Comment (tgraf) [reply bpf/lib/policy.h:73]: Makes sense

**Review by tgraf — APPROVED** [review]

**Comment by joestringer** [comment]
test-me-please

**Comment by tgraf** [comment]
Legit CI failure:
```
12:35:25     runtime:         if (HAVE_L4_POLICY && !is_fragment) {
12:35:25     runtime:             ^~~~~~~~~~~~~~
12:35:25     runtime:             DBG_L4_POLICY
12:35:25     runtime: ./lib/dbg.h:64:2: note: 'DBG_L4_POLICY' declared here
12:35:25     runtime:         DBG_L4_POLICY,
12:35:25     runtime:         ^
12:35:25     runtime: In file included from bpf_netdev.c:46:
12:35:25     runtime: ./lib/policy.h:93:6: error: use of undeclared identifier 'HAVE_L4_POLICY'; did you mean 'DBG_L4_POLICY'?
12:35:25     runtime:         if (HAVE_L4_POLICY && !is_fragment) {
12:35:25     runtime:             ^~~~~~~~~~~~~~
12:35:25     runtime:             DBG_L4_POLICY
12:35:25     runtime: ./lib/dbg.h:64:2: note: 'DBG_L4_POLICY' declared here
12:35:25     runtime:         DBG_L4_POLICY,
12:35:25     runtime:         ^
12:35:25     runtime: In file included from bpf_netdev.c:46:
12:35:25     runtime: ./lib/policy.h:78:9: error: use of undeclared label 'get_proxy_port'
12:35:25     runtime:                         goto get_proxy_port;
12:35:25     runtime:                              ^
```

I'm also fine to squash this with #5332 as the faulty code is removed with that PR

**Comment by ianvernon** [comment]
@tgraf @joestringer given that #5332 has the commit from this PR included within it, and that build passed, can we close this one and use that PR instead? That would resolve the build failure.

**Comment by tgraf** [comment]
@ianvernon It would break bisecting so I'd rather rebase it properly.

**Comment by ianvernon** [comment]
@tgraf fair, makes sense.

**Comment by joestringer** [comment]
Should be fixed now, forgot that this was why I had the `#ifndef` bit previously.

**Comment by joestringer** [comment]
test-me-please

**Comment by ianvernon** [comment]
Merging: 
https://github.com/cilium/cilium/pull/5332 was approved by @borkmann , which includes the code from this PR, so this is effectively received @cilium/bpf approval.

**Pull request merged**